File: CodeArduinoRuche/TxLoRaE5.cpp
#include "TxLoRaE5.h"

//-----------------------------------------------Fonction-----------------------------------------------
void initialisationLoRa_E5_P2P(uint16_t frequence,  String SpreadFactor, uint16_t BandWidth, uint8_t TX_Preamble, uint8_t RX_Preamble, uint8_t Power, String CRC, String Inverted_IQ, String Public_LoRaWAN, Stream &portSerieVirtuel, uint8_t *initialisaitonLoRa)
{
  /*
    Fonction d initialisation du module LoRa-E5
    
    Args:
      frequence (uint16_t): frequence d emission du module
      SpreadFactor (String): facteur d empietement
      BandWidth (uint16_t): bande passante 
      TX_Preamble (uint8_t): longeur des datas en emission
      RX_Preamble (uint8_t): longeur des datas en reception
      Power (uint8_t): puissance d emission en dBm
      CRC (String): activation du CRC (activation du CRC "ON")
      Inverted_IQ (String): sens du changement de frequence de la modulation
      Public_LoRaWAN (string): si on veut oui ou non utiliser le reseau LoRaWan
      portSerieVirtuel (Stream): nom de la class du port serie physique du module LoRa
      initialisaitonLoRa (uint8_t) variable d acquittement de la bonne initialisation du module LoRa. si TRUE erreur de l init
  */

  //variable local
  String LoRaRecu = "";
  uint16_t timeOut = 0; 

  *initialisaitonLoRa = 0;
  
  //passage du module Lora en "mode TEST"
  if(*initialisaitonLoRa ==0){
    portSerieVirtuel.print("AT+MODE=TEST");
    timeOut = 0;
    while(portSerieVirtuel.available() == 0 & timeOut < 5000)
    {
      delay(10);
      timeOut = timeOut +10;
    }
    LoRaRecu = portSerieVirtuel.readString();
    if(LoRaRecu.indexOf("+MODE: TEST") <= 0)
    {
      //passage de la variable d initialisation a false et ecriture d un message d erreur
      *initialisaitonLoRa = 1; 
    }
  }
  
  //passage du module Lora en "mode STOP" (empeche le module de recevoir des information durant la reprogramation du mode)
  if(*initialisaitonLoRa ==0){
    portSerieVirtuel.print("AT+TEST=STOP");
    timeOut = 0;
    while(portSerieVirtuel.available() == 0 & timeOut < 5000)
    {
      delay(10);
      timeOut = timeOut +10;
    }
    LoRaRecu = portSerieVirtuel.readString();
    if(LoRaRecu.indexOf("+TEST: STOP") <= 0)
    {
      *initialisaitonLoRa = 1;
    }
  }

  //reglage du LDRO du module Lora
  if(*initialisaitonLoRa ==0){
    portSerieVirtuel.print("AT+LW=LDRO,ON");
    timeOut = 0;
    while(portSerieVirtuel.available() <= 0 & timeOut < 5000)
    {
      delay(10);
      timeOut = timeOut +10;
    }
    LoRaRecu = portSerieVirtuel.readString();
    if(LoRaRecu.indexOf("+LW: LDRO, ON") <= 0)
    {
      *initialisaitonLoRa = 1;
    }
  }

  //configuration des parametres d emission radio du module LoRa
  if(*initialisaitonLoRa ==0){
    portSerieVirtuel.print("AT+TEST=RFCFG," + String(frequence) + "," + SpreadFactor + "," + String(BandWidth) + "," + String(TX_Preamble) + "," + String(RX_Preamble) + "," + String(Power) + "," + CRC + "," + Inverted_IQ + "," + Public_LoRaWAN);
    timeOut = 0;
    while(portSerieVirtuel.available() <= 0 & timeOut < 5000)
    {
      delay(10);
      timeOut = timeOut + 10;
    }
    LoRaRecu = portSerieVirtuel.readString();
    if(LoRaRecu.indexOf("RFCFG F") <=0)
    {
      *initialisaitonLoRa = 1;
    }
  }  
}


void envoieLoRa_E5_P2P(uint16_t cleSysteme, uint8_t idRuche, uint8_t typeCapteur, float valeurCapteur, Stream &portSerieVirtuel, uint8_t *receptionLoRa)
{
  /*
    Fonction utiliser pour envoyer une donnee en peer to peer entre deux module Lora-E5
    
    Args:
      cleSysteme (uint16_t): cle utilisee pour dissocier les messages du projet ruche des autres messages sur la frequence 868/433Mhz
      idRuche (uint8_t): numerotation de la ruche, utiliser pour dissocier plusieur ruche entre elle sur la meme frequence
      typeCapteur (uint8_t): numerotaion du capteur utiliser ie savoir si c est un capteur de temperature ou de poid etc...
      valeurCapteur (float): valeur lu par le capteur 
      portSerieVirtuel (Stream): nom de la class du port serie physique du module LoRa
      receptionLoRa (uint8_t) flag de la bonne emission d un message LoRa. si TRUE erreur de l emission du message
  */
  
  //structure de convertion des float en uin32_t (representation ieee754 des floatants)
  typedef union
  {
    float asFloat;
    uint32_t asUint32_t;
  }
  convertStruct;

  //declaration des variables locals
  convertStruct structValue;
  String StringCleSysteme ="";
  String StringIdRuche ="";
  String StringTypeCapteur ="";
  String StringValeurCapteur ="";  
  String LoRaRecu ="";
  uint16_t timeOut = 0;
  //pointeur flag d erreur
  *receptionLoRa = 0;

  //convertie la valeur de valeurCapteur en hex sur 32bits (ieee754)
  structValue.asFloat = valeurCapteur;
  StringValeurCapteur = String(structValue.asUint32_t, HEX);
  while(StringValeurCapteur.length() < 8)
  {
    StringValeurCapteur = "0" + StringValeurCapteur;
  }

  //convertie la valeur de cleSysteme en hex sur 16bits
  StringCleSysteme = String(cleSysteme, HEX);
  while(StringCleSysteme.length() < 4)
  {
    StringCleSysteme = "0" + StringCleSysteme;
  }

  //convertie la valeur de l'idRuche en hex sur 8bits
  StringIdRuche = String(idRuche, HEX);
  while(StringIdRuche.length() < 2)
  {
    StringIdRuche = "0" + StringIdRuche;
  }

  //convertie la valeur du typeCapteur en hex sur 8bits
  StringTypeCapteur = String(typeCapteur, HEX);
  while(StringTypeCapteur.length() < 2)
  {
    StringTypeCapteur = "0" + StringTypeCapteur;
  }

  //envoie des donnee via le module LoRa
  portSerieVirtuel.print("AT+TEST=TXLRPKT, " + StringCleSysteme + StringIdRuche + StringTypeCapteur + StringValeurCapteur);

  //verification du bon envoie des donnes
  timeOut = 0;
  LoRaRecu = "";
  while(LoRaRecu.indexOf("+TEST: TX DONE") <= 0 & timeOut < 5000)
  {
    //concatenation des value renvoyer par le module LoRa
    if(portSerieVirtuel.available() != 0){
      LoRaRecu = LoRaRecu + portSerieVirtuel.readString();
    }
    //gestion du timeout
    delay(10);
    timeOut = timeOut +10;
  }
  if(LoRaRecu.indexOf("+TEST: TX DONE") <= 0)
  {
    *receptionLoRa = 1;
  }
}

File: ExempleCode/RasberryPi/P2P_LoRa-E5/TxLoRa-E5.py
#importation des libraries utilisees
import serial #pyserial
import time 
import sys


#----------Autre fonction----------
def lectureLoRa(port, octetLecture=200, stringAttendu="", timeOut=5):
    """Fonction de lecture serie des information retourner par le module LoRa-E5
        si les octets rendu par la carte LoRa ne correspondent pas a "stringAttendu" alors la fonction atteindra un timeout et rendra la valeur "-1"

    Args:
        port (class): class utiliser pour declarer le port serie
        octetLecture (int): nombre d octet a lire a la fois
        stringAttendu (str): string que le module LoRa doit normalement nous rendre
        timeOut (int): valeur du timeout a attendre si l on ne recois pas le message "stringAttendu"
    """

    octetRecu = ""
    timeout_start = time.time()
    while((octetRecu.find(stringAttendu) == -1) & (time.time() < timeout_start + timeOut)):
        while(port.inWaiting() < 1) & (time.time() < timeout_start + timeOut): 
            time.sleep(0.25)
        while(1):
            octetLu = port.read(size=octetLecture).decode("utf-8")
            octetRecu = octetRecu + octetLu
            if len(octetLu) != octetLecture:
                break
    if(octetRecu.find(stringAttendu) == -1):
        octetRecu = "-1"
    return(octetRecu)


def initLoRa(frequence, SpreadFactor, BandWidth, TX_Preamble, RX_Preamble, Power, CRC, Inverted_IQ, Public_LoRaWAN, portSerieLoRa):
    """Fonction d initialisation du module LoRa E5 en emeteur

    Args:
        frequence (int): frequence d emission du module
        SpreadFactor (string): bande passante du module
        BandWidth (int): bande passante 
        TX_Preamble (int):
        RX_Preamble (int): 
        Power (int): puissance d emission en dBm
        CRC (string): activation du CRC (activation du CRC "ON")
        Inverted_IQ (string): 
        Public_LoRaWAN (string): action ou non du module en mode privee ou public
        portSerieLoRa (string): nom de la class du port serie physique du module LoRa
        
    Returns:
        int: 1 ou 0 selon si l initialisation c est bien passer ou non
    """

    #variable local
    initialisaitonLoRaOK = 1
    strAEnvoier = ""

    #passage du module Lora en "mode TEST"
    portSerieLoRa.write('AT+MODE=TEST'.encode())
    receptionLoRa = lectureLoRa(port=portSerieLoRa, octetLecture=200, stringAttendu="+MODE: TEST", timeOut=5)
    if(receptionLoRa =="-1"):
        initialisaitonLoRaOK = 0
        print("\n erreur ecriture mode TEST")
    else:
        print(receptionLoRa)

    #passage du module Lora en "mode STOP" (empeche le module de recevoir des information durant la reprogramation du mode)
    if(initialisaitonLoRaOK == 1):
        portSerieLoRa.write('AT+TEST=STOP'.encode())
        receptionLoRa = lectureLoRa(port=portSerieLoRa, octetLecture=200, stringAttendu="+TEST: STOP", timeOut=5)
        if(receptionLoRa =="-1"):
            initialisaitonLoRaOK = 0
            print("\n erreur ecriture mode STOP")
        else:
            print(receptionLoRa)

    #reglage du LDRO du module Lora
    if(initialisaitonLoRaOK == 1):
        portSerieLoRa.write('AT+LW=LDRO,ON'.encode())
        receptionLoRa = lectureLoRa(port=portSerieLoRa, octetLecture=200, stringAttendu="+LW: LDRO, ON", timeOut=5)
        if(receptionLoRa =="-1"):
            initialisaitonLoRaOK = 0
            print("\n erreur ecriture mode LDRO")
        else:
            print(receptionLoRa)

    #configuration des parametres d emission radio du module LoRa
    if(initialisaitonLoRaOK == 1):
        strAEnvoier = 'AT+TEST=RFCFG,' + str(frequence) + ',' + SpreadFactor + ',' + str(BandWidth) + ',' + str(TX_Preamble) + ',' + str(RX_Preamble) + ',' + str(Power) + ',' + CRC + ',' + Inverted_IQ + ',' + Public_LoRaWAN
        portSerieLoRa.write(strAEnvoier.encode())
        receptionLoRa = lectureLoRa(portSerieLoRa, octetLecture=200, stringAttendu="+TEST: RFCFG F", timeOut=5)
        if(receptionLoRa =="-1"):
            initialisaitonLoRaOK = 0
            print("\n erreur ecriture mode RFCFG")
        else:
            print(receptionLoRa)

    return initialisaitonLoRaOK


def emission(portSerieLoRa, codeSysteme, codeRuche, indiceCapteur, valeurCapteur):
    """Fonction d emission d un message LoRa avec verification de la bonne emission d un message

    Args:
        portSerieLoRa (string): nom de la class du port serie physique du module LoRa
        codeSysteme (int): mot de passe utiliser pour la ruche
        codeRuche (int): indice de la ruche
        indiceCapteur (int): indice du capteur de la ruche
        valeurCapteur (uint32bits (ieee754)): valeur du capteur a envoyer
        
    Returns:
        int: 1 ou 0 selon si l emission du message c est bien passer ou non
    """

    #variable local
    emissionLoRaOK = 1
    strCodeSysteme = ""
    strcodeRuche = ""
    strindiceCapteur = "" 
    strvaleurCapteur = ""

    #prise en compte de la taille valeurCapteur
    strvaleurCapteur = hex(valeurCapteur)[2:]
    while (len(strvaleurCapteur) < 8):
        strvaleurCapteur = "0" + strvaleurCapteur

    #prise en compte de la taille du codeSysteme
    strCodeSysteme = hex(codeSysteme)[2:]
    while (len(strCodeSysteme) < 4):
        strCodeSysteme = "0" + strCodeSysteme
       
    #prise en compte de la taille du codeRuche
    strcodeRuche = hex(codeRuche)[2:]
    while (len(strcodeRuche) < 2):
        strcodeRuche = "0" + strcodeRuche

    #prise en compte de la taille du indiceCapteur
    strindiceCapteur = hex(indiceCapteur)[2:]
    while (len(strindiceCapteur) < 2):
        strindiceCapteur = "0" + strindiceCapteur

    #preparation et envoie de la trame 
    LoRaDataAEnvoyer = 'AT+TEST=TXLRPKT, "' + strCodeSysteme + strcodeRuche + strindiceCapteur + strvaleurCapteur + '"' 
    portSerieLoRa.write(LoRaDataAEnvoyer.encode())

    #reception de l'acquittement du module LoRa 
    receptionLoRa = lectureLoRa(portSerieLoRa, octetLecture=200, stringAttendu="TX DONE", timeOut=5)
    if(receptionLoRa =="-1"):
        emissionLoRaOK = 0
    else:
        print(receptionLoRa)

    return emissionLoRaOK


#----------Fonction main----------
def main():
    #initialisation du port serie
    ser = serial.Serial(
        port='/dev/ttyAMA0',
        baudrate=9600,
        bytesize = serial.EIGHTBITS,
        stopbits = serial.STOPBITS_ONE,
        parity = serial.PARITY_NONE,
        xonxoff = True,
        timeout=0.1
    )

    #verification de l initialisation de l UART
    if(ser.isOpen()):
        ser.flushInput()
        ser.flushOutput()

        #initialisation du module LoRa
        resulInitLoRa = initLoRa(
            frequence = 868, 
            SpreadFactor = "SF12", 
            BandWidth = 125, 
            TX_Preamble = 12, 
            RX_Preamble = 15, 
            Power = 14, 
            CRC = "ON", 
            Inverted_IQ = "OFF", 
            Public_LoRaWAN = "OFF", 
            portSerieLoRa =ser)

        #gestion du resultat de l initialisation
        if(resulInitLoRa != 1):
            print("intialisation du module LoRa KO")
            print("fermeture du programme")
            sys.exit()
        print("initialisation du module LoRa OK")

        #envoie la meme data en continue pour tester la reception des datas en mode P2P
        while(True):
            #emission d un message LoRa
            resulEmissionLoRa = emission(portSerieLoRa = ser,
                codeSysteme = 0x2C33,
                codeRuche = 0x00,
                indiceCapteur = 0x1,
                valeurCapteur = 0xc7f1207e)

            #gestion de l erreur demission
            if(resulEmissionLoRa !=1):
                print("Erreur emission message LoRa")
                print("fermeture du programme")
                sys.exit()

            #delay d attente
            time.sleep(10)
    else:
        print("Erreur initialisation du module UART")
        print("fermeture du programme")
        sys.exit()


#lancement de la fonction main
if __name__ == '__main__':
    main()